addons/crm/static/src/js/forecast/forecast_service.js

```diff
@@ -0,0 +1,230 @@
+/** @odoo-module alias=crm.ForecastService **/
+import AbstractService from 'web.AbstractService';
+import core from 'web.core';
+
+/**
+ * File intended for `web` (can be used in all kind of forecast configurations)
+ * 
+ * Forecast Period:
+ *   Represents a specific date/time range for a specific model, field and granularity.
+ * 
+ * It is used to add new domain and context constraints related to a specific date/time
+ * field, in order to configure the _read_group_fill_temporal (see core models.py)
+ * method. It will be used when we want to get continuous groups in chronological
+ * order in a specific date/time range.
+ */
+
+/**
+ * Configuration depending on the granularity:
+ * startOf : method to get the start moment of the period from a moment
+ * cycle : amount of 'granularity' periods constituting a cycle
+ * cyclePos : method to get the position (index) in the cycle from a moment (1 is the first index)
```

Comment:
`moment.quarter` and `moment.isoWeekday` start at index 1, but `moment.hour` and `moment.month` start at index 0. The `+1` is to standardize the first index, because I use the same formula (`computeEnd`) for each granularity ^^

It is just an index which has a maximal value of `cycle`.
The `cycle` duration is arbitrary: I chose 
* 1 day as the cycle for `hour`
* 1 week as the cycle for `day` and `week` -> because there is no greater time period that takes an integer number of weeks
* 1 year as the cycle for `month`, `quarter` and `year`

I'll add a comment for this ^^

Reply:
Ah ok, indexes makes sense. Indeed deserves a comment with that explanation :)